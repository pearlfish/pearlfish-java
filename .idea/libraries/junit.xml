<component name="libraryTable">
  <library name="junit">
    <CLASSES>
      <root url="jar://$PROJECT_DIR$/libs/junit/junit-4.11.jar!/" />
    </CLASSES>
    <JAVADOC />
    <SOURCES>
      <root url="jar://$PROJECT_DIR$/libs/junit/junit-4.11-sources.jar!/" />
    </SOURCES>
  </library>
</component>