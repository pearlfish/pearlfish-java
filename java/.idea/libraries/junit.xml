<component name="libraryTable">
  <library name="junit">
    <CLASSES>
      <root url="file://$PROJECT_DIR$/libs/junit" />
    </CLASSES>
    <JAVADOC />
    <SOURCES>
      <root url="file://$PROJECT_DIR$/libs/junit" />
    </SOURCES>
    <jarDirectory url="file://$PROJECT_DIR$/libs/junit" recursive="false" />
    <jarDirectory url="file://$PROJECT_DIR$/libs/junit" recursive="false" type="SOURCES" />
  </library>
</component>