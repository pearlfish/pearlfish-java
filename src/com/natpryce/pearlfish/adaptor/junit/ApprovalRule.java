package com.natpryce.pearlfish.adaptor.junit;

import com.natpryce.pearlfish.ApprovalError;
import com.natpryce.pearlfish.Approver;
import com.natpryce.pearlfish.FileNamingConvention;
import com.natpryce.pearlfish.TestSpecificFormat;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

import java.io.IOException;

public class ApprovalRule<T> implements TestRule {
    private final TestSpecificFormat<? super T> format;
    private final FileNamingConvention namingConvention;

    private Approver<T> approver = null;

    public ApprovalRule(TestSpecificFormat<? super T> format, FileNamingConvention namingConvention) {
        this.namingConvention = namingConvention;
        this.format = format;
    }

    public void check(T receivedContents) throws IOException {
        try {
            approver().check(receivedContents);
        } catch (ApprovalError e) {
            throw new ApprovalFailure(e);
        }
    }

    public void recordAsApproved(T receivedContents) throws IOException {
        approver().recordAsApproved(receivedContents);
    }

    private Approver<T> approver() {
        if (approver == null) {
            throw new IllegalStateException("cannot perform approval before discovering the name of the test (did you forget to annotate the approver with @Rule?)");
        }

        return approver;
    }

    @Override
    public Statement apply(Statement base, Description description) {
        Class<?> testClass = description.getTestClass();
        String testName = description.getMethodName();
        approver = new Approver<T>(
                testClass,
                testName,
                namingConvention,
                format.forTest(testClass, testName));
        return base;
    }
}
