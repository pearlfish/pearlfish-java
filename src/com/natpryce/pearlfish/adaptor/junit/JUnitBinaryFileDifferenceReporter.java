package com.natpryce.pearlfish.adaptor.junit;

import com.natpryce.pearlfish.internal.IO;

import java.io.File;
import java.io.IOException;

public class JUnitBinaryFileDifferenceReporter extends JUnitDifferenceReporter {
    @Override
    protected String loadContents(File file) throws IOException {
        return IO.toHexDump(file);
    }
}
