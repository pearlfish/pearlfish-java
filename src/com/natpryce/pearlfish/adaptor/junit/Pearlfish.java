package com.natpryce.pearlfish.adaptor.junit;

import com.natpryce.pearlfish.TestSpecificFormat;
import com.natpryce.pearlfish.formats.Formats;
import com.natpryce.pearlfish.naming.NextToSourceNamingConvention;

import java.io.File;

public class Pearlfish {
    public static <T> ApprovalRule<T> approvalRule(final String srcRoot) {
        return approvalRule(srcRoot, Formats.PLAIN_TEXT);
    }

    public static <T> ApprovalRule<T> approvalRule(final String srcRoot, final TestSpecificFormat<? super T> format) {
        return new ApprovalRule<T>(format, new NextToSourceNamingConvention(new File(srcRoot)));
    }
}
