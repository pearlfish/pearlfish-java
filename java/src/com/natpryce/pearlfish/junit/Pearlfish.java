package com.natpryce.pearlfish.junit;

import com.natpryce.pearlfish.formats.MarkdownFormatter;
import org.rococoa.okeydoke.Approver;
import org.rococoa.okeydoke.ApproverFactory;
import org.rococoa.okeydoke.Reporters;
import org.rococoa.okeydoke.junit.ApprovalsRule;
import org.rococoa.okeydoke.sources.FileSystemSourceOfApproval;

import java.io.File;
import java.nio.charset.Charset;

public class Pearlfish {
    public static ApprovalsRule pearlfishApprovalsRule(final String srcRoot) {
        return new ApprovalsRule(new ApproverFactory() {
            public Approver create(String testName, Class<?> testClass) {
                MarkdownFormatter formatter = new MarkdownFormatter(testClass, testName, Charset.defaultCharset());

                FileSystemSourceOfApproval sourceOfApproval = new FileSystemSourceOfApproval(
                        new File(srcRoot), testClass.getPackage()).
                        withTypeExtension(formatter.fileExtension());

                return new Approver(testName, sourceOfApproval, formatter, Reporters.reporter());
            }
        });
    }
}
