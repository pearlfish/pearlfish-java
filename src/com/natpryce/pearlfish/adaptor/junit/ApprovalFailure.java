package com.natpryce.pearlfish.adaptor.junit;

import org.junit.ComparisonFailure;

import java.io.File;

public class ApprovalFailure extends ComparisonFailure {
    public final File approvedFile;
    public final String approvedContents; // empty string if approvedFile does not exist.
    public final File receivedFile;
    public final String receivedContents;

    public ApprovalFailure(String message, File approvedFile, String approvedContents, File receivedFile, String receivedContents) {
        super(message, approvedContents, receivedContents);

        this.approvedFile = approvedFile;
        this.approvedContents = approvedContents;
        this.receivedFile = receivedFile;
        this.receivedContents = receivedContents;
    }
}
