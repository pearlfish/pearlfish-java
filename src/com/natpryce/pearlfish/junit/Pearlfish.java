package com.natpryce.pearlfish.junit;

import com.natpryce.pearlfish.formats.Format;
import com.natpryce.pearlfish.formats.Formats;
import com.natpryce.pearlfish.formats.MissingTemplateException;
import org.rococoa.okeydoke.Approver;
import org.rococoa.okeydoke.ApproverFactory;
import org.rococoa.okeydoke.Formatter;
import org.rococoa.okeydoke.Reporters;
import org.rococoa.okeydoke.junit.ApprovalsRule;
import org.rococoa.okeydoke.sources.FileSystemSourceOfApproval;

import java.io.File;

public class Pearlfish implements Formats {
    public static ApprovalsRule approvalRule(final String srcRoot, final Format format) {
        return new ApprovalsRule(new PearlfishApproverFactory(srcRoot, format));
    }

    private static class PearlfishApproverFactory implements ApproverFactory {
        private final String srcRoot;
        private final Format preferredFormat;

        public PearlfishApproverFactory(String srcRoot, Format preferredFormat) {
            this.srcRoot = srcRoot;
            this.preferredFormat = preferredFormat;
        }

        public Approver create(String testName, Class<?> testClass) {
            Formatter<Object,String> formatter;
            String fileExtension;
            try {
                formatter = preferredFormat.formatterFor(testName, testClass);
                fileExtension = preferredFormat.fileExtension();
            }
            catch (MissingTemplateException e) {
                formatter = YAML.formatterFor(testName, testClass);
                fileExtension = YAML.fileExtension();
            }

            return new Approver(
                    testName,
                    new FileSystemSourceOfApproval(
                            new File(srcRoot), testClass.getPackage()).
                            withTypeExtension(fileExtension),
                    formatter,
                    Reporters.reporter());
        }
    }
}
