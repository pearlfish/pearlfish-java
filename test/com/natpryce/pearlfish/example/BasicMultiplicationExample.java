package com.natpryce.pearlfish.example;


import com.natpryce.pearlfish.junit.Pearlfish;
import org.junit.Rule;
import org.junit.Test;
import org.rococoa.okeydoke.junit.ApprovalsRule;

import java.io.IOException;
import java.math.BigInteger;

import static com.natpryce.pearlfish.Results.*;
import static com.natpryce.pearlfish.formats.Formats.MARKDOWN;

public class BasicMultiplicationExample {
    public @Rule ApprovalsRule approvals = Pearlfish.approvalRule("test", MARKDOWN);

    @Test
    @SuppressWarnings("unchecked")
    public void multiplication() throws IOException {
        approvals.assertApproved(results(
                section("basic",
                        scenario("simple multiplication", new Operands(2, 5), BigInteger.valueOf(10)),
                        scenario("zero left", new Operands(0, 2), BigInteger.valueOf(0)),
                        scenario("zero right", new Operands(1, 0), BigInteger.valueOf(0)),
                        scenario("zero both", new Operands(0, 0), BigInteger.valueOf(0))),
                section("negative",
                        scenario("negative left", new Operands(-4, 2), BigInteger.valueOf(-8)),
                        scenario("negative right", new Operands(5, -4), BigInteger.valueOf(-20)),
                        scenario("negative left and zero", new Operands(-1, 0), BigInteger.valueOf(0)),
                        scenario("zero and negative right", new Operands(0, -6), BigInteger.valueOf(0)),
                        scenario("both negative", new Operands(-4, -9), BigInteger.valueOf(36))),
                section("large",
                        scenario("large addition", new Operands(Integer.MAX_VALUE, 2), BigInteger.valueOf(Integer.MAX_VALUE).multiply(BigInteger.valueOf(2))),
                        scenario("large negative numbers", new Operands(Integer.MIN_VALUE, 2), BigInteger.valueOf(Integer.MIN_VALUE).multiply(BigInteger.valueOf(2))))));
    }

}
