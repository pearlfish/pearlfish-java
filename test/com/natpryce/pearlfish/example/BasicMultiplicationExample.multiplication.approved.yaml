basic:
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: 2, y: 5}
  name: simple multiplication
  output: 10
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: 0, y: 2}
  name: zero left
  output: 0
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: 1, y: 0}
  name: zero right
  output: 0
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: 0, y: 0}
  name: zero both
  output: 0
negative:
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: -4, y: 2}
  name: negative left
  output: -8
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: 5, y: -4}
  name: negative right
  output: -20
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: -1, y: 0}
  name: negative left and zero
  output: 0
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: 0, y: -6}
  name: zero and negative right
  output: 0
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: -4, y: -9}
  name: both negative
  output: 36
large:
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: 2147483647, y: 2}
  name: large addition
  output: 4294967294
- !!com.natpryce.pearlfish.Scenario
  input: !!com.natpryce.pearlfish.example.Operands {x: -2147483648, y: 2}
  name: large negative numbers
  output: -4294967296
